osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,202 @@
+            {
+                foreach (var channel in leftChannels)
+                    channel.NewMessagesArrived -= channel_NewMessagesArrived;
+            };
+        }
+
+        private void channel_NewMessagesArrived(IEnumerable<Message> messages)
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        /// <remarks>
+        /// Resolves the channel id
+        /// </remarks>
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.SingleOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // Only send notifications, if ChatOverlay and the target channel aren't visible.
+            if (chatOverlay?.IsPresent == true && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages.OrderByDescending(m => m.Id))
+            {
+                // ignore messages that already have been read
+                if (message.Id <= channel.LastReadId)
+                    return;
+
+                if (message.Sender.Id == localUser.Value.Id)
+                    continue;
+
+                // check for private messages first, if true, skip checking mentions to prevent duplicate notifications about the same message.
+                if (checkForPMs(channel, message))
+                    continue;
+
+                // change output to bool again if another "message processor" is added.
+                checkForMentions(channel, message, localUser.Value.Username);
+            }
+        }
+
+        private bool checkForPMs(Channel channel, Message message)
+        {
+            if (!notifyOnChat.Value || channel.Type != ChannelType.PM)
+                return false;
+
+            var notification = new PrivateMessageNotification(message.Sender.Username, channel);
+
+            notificationOverlay?.Post(notification);
+
+            return true;
+        }
+
+        private void checkForMentions(Channel channel, Message message, string username)
+        {
+            if (!notifyOnMention.Value || !anyCaseInsensitive(getWords(message.Content), username))
+                return;
+
+            var notification = new MentionNotification(message.Sender.Username, channel);
+            notificationOverlay?.Post(notification);
+        }
+
+        private IEnumerable<string> getWords(string input) => Regex.Matches(input, @"\w+", RegexOptions.Compiled).Select(c => c.Value);
```

Comment:
Why a regex and not `string.Split(' ')`? This will fail in the case of names like `[John]` or `Doe-`, or any other non-A/N characters we support.

Reply:
I initially proposed to change from splitting over whitespace because of punctuation. You're right that it will fail for search strings with non-word characters; should go back to splitting over whitespace, but check search term containment without case sensitivity.

Reply:
The other way to do this is not splitting, but by searching the string, which would cover the punctuation case. Stable already does this and I think it's worked well?

Also: need to replace any space in the name with `_`.

Reply:
I will remove the word checking, but are you sure with the underscore (or is it stable behavior)?

Reply:
It's stable behaviour. Upon further discussion, do both (match on both no-underscores and with-underscores).